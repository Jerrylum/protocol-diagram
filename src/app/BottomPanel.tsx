import { Box } from "@mui/material";
import { observer } from "mobx-react-lite";
import { CommandInputField } from "./CommandInputField";

export const BottomPanel = observer(() => {
  return (
    <Box
      sx={{
        position: "fixed",
        bottom: "32px",
        maxWidth: "90vw",
        width: "600px",
        left: "50%",
        transform: "translate(-50%, 0)"
      }}>
      <CommandInputField />
    </Box>
  );
});
