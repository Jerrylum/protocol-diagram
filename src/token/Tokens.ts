// ALGO: Tokens implementation is adopted from https://github.com/Jerrylum/ProtocolDiagram under the GPLv3 license.

import isEqual from "lodash.isequal";

/**
 * A utility function that checks whether the character is delimiter (null | ' ')
 *
 * @param c the character to be checked
 * @return whether the character is delimiter
 */
export function isDelimiter(c: string | null): boolean {
  return c === null || c === " ";
}

/**
 * A utility function that checks whether the character is delimiter (null | ' ' | ':' | ',')
 *
 * @param c the character to be checked
 * @return whether the character is delimiter
 */
export function isSafeDelimiter(c: string | null): boolean {
  return (
    c === null ||
    c === " " ||
    c === ":" ||
    c === "," ||
    c === "+" ||
    c === "-" ||
    c === "*" ||
    c === "/" ||
    c === "(" ||
    c === ")"
  );
}

/**
 * A utility function that reads the code point buffer, returns a Token instance
 * typed with T which given by the parameter class T if the code point buffer
 * does match one of our specified acceptable characters, else return null.
 *
 * @template T - The type of the token to be created.
 * @param {CodePointBuffer} buffer - The code point buffer to be read.
 * @param {string[]} accepts - The array of acceptable characters.
 * @param {new (value: string) => T} tokenClass - The constructor of the token to be created.
 * @returns {T|null} - The token instance, or null if the buffer does not match any acceptable characters.
 */
function doParseCodepoint<T extends Token>(
  buffer: CodePointBuffer,
  accepts: string[],
  tokenClass: new (value: string) => T
): T | null {
  buffer.savepoint();

  const target = buffer.next();
  for (const c of accepts) {
    if (target === c) {
      return buffer.commitAndReturn(new tokenClass(c));
    }
  }
  return buffer.rollbackAndReturn(null);
}

/**
 * Parses a quoted string token from the given code point buffer.
 * @template T - The type of the token.
 * @param {CodePointBuffer} buffer - The code point buffer to be parsed.
 * @param {string} quote - The quote character.
 * @param {new (value: string, content: string) => T} tokenClass - The class of the token to be returned.
 * @returns {T|null} - The parsed token instance or null if the buffer does not contain a valid quoted string.
 */
function doParseQuoteString<T extends Token>(
  buffer: CodePointBuffer,
  quote: string,
  tokenClass: new (value: string, content: string) => T
): T | null {
  buffer.savepoint();

  let valueSb = "";
  let contentSb = "";
  let escape = false;
  let open = true;

  if (buffer.next() !== quote) {
    return buffer.rollbackAndReturn(null);
  }
  valueSb += quote;

  while (buffer.hasNext()) {
    const c = buffer.next();

    if (escape) {
      escape = false;
      contentSb += c;
      valueSb += c;
    } else {
      if (c === "\\") {
        escape = true;
        valueSb += c;
      } else if (c === quote) {
        open = false;
        valueSb += c;
        break;
      } else {
        valueSb += c;
        contentSb += c;
      }
    }
  }

  if (open) {
    return buffer.rollbackAndReturn(null);
  } else {
    return buffer.commitAndReturn(new tokenClass(valueSb, contentSb));
  }
}

/**
 * A class that wraps a string for the ease of tokenizing and parsing.
 */
export class CodePointBuffer {
  /**
   * The be-wrapped string.
   */
  private target: string;
  /**
   * The character-reading cursor position.
   */
  private index: number;
  /**
   * The cursor position history, will be used to rewind to historic positions during the parsing.
   */
  private history: number[];

  /**
   * @param target The string to be wrapped.
   */
  constructor(target: string) {
    this.target = target;
    this.index = 0;
    this.history = [];
  }

  /**
   * A method to retrieve the length of the internally stored string.
   */
  public length(): number {
    return this.target.length;
  }

  /**
   * A method to get the current cursor index.
   */
  public getIndex(): number {
    return this.index;
  }

  /**
   * A method to lookup the character by specified index.
   * @param index The index to be looked up.
   * @returns The character at the specified index, or null if the index is out of range.
   */
  public at(index: number): string | null {
    return index < this.length() ? this.target.charAt(index) : null;
  }

  /**
   * A method to save the current cursor into the history.
   */
  public savepoint(): void {
    this.history.push(this.index);
  }

  /**
   * A method to retrieve the last saved cursor position, and restore the cursor from that position.
   */
  public rollback(): void {
    this.index = this.history.pop()!;
  }

  /**
   * A wrapper function that receives and returns the value without performing any operation on top of that value and rollback behind the scene.
   * @template T
   * @param value The value to be wrapped.
   * @returns The value that was passed in.
   */
  public rollbackAndReturn<T>(value: T): T {
    this.rollback();
    return value;
  }

  /**
   * A method that remove the last saved cursor.
   */
  public commit(): void {
    this.history.pop();
  }

  /**
   * A wrapper function that receives and returns the value without performing any operation on top of that value and commit behind the scene.
   * @template T
   * @param value The value to be returned.
   * @returns The value that was passed in.
   */
  public commitAndReturn<T>(value: T): T {
    this.commit();
    return value;
  }

  /**
   * A method to lookup the cursor pointing character and move the cursor to the right.
   * @returns The character at the current cursor position, or null if the cursor is at the end of the string.
   */
  public next(): string | null {
    return this.at(this.index++);
  }

  /**
   * A method to lookup the cursor pointing character.
   * @returns The character at the current cursor position, or null if the cursor is at the end of the string.
   */
  public peek(): string | null;

  /**
   * A method to look a head character by given offset.
   * @param offset The offset to look ahead.
   * @returns The character at the specified offset ahead of the current cursor position, or null if the offset is out of range.
   */
  public peek(offset: number): string | null;

  public peek(offset?: number): string | null {
    return this.at(offset !== undefined ? this.index + offset : this.index);
  }

  /**
   * A method to determine whether the string could be further consumed.
   * @returns true if there are more characters to consume, false otherwise.
   */
  public hasNext(): boolean {
    return this.index < this.length();
  }

  /**
   * A method that reads every delimiter and stops once it bumps into a non-delimiter character and return.
   * @returns The number of delimiters that were read.
   */
  public readDelimiter(): number {
    let count = 0;
    while (this.hasNext() && isDelimiter(this.peek())) {
      count++;
      this.next();
    }
    return count;
  }

  /**
   * A method that reads every character and stops once it bumps into a delimiter.
   * @returns A string containing all characters read up to the delimiter, or an empty string if the cursor is at a delimiter.
   */
  public readChunk(): string {
    let rtn = "";
    while (this.hasNext() && !isDelimiter(this.peek())) {
      rtn += this.next();
    }
    return rtn; // might be empty
  }

  /**
   * A method that reads every character and stops once it bumps into a safe delimiter, which is a superset of delimiter, it contains (null | ' ' | ':' | ',' and more).
   * @returns A string containing all characters read upto the safe delimiter, or an empty string if the cursor is at a safe delimiter.
   */
  public readSafeChunk(): string {
    let rtn = "";
    while (this.hasNext() && !isSafeDelimiter(this.peek())) {
      rtn += this.next();
    }
    return rtn; // might be empty
  }
}

export abstract class Token {
  public static parse(buffer: CodePointBuffer): Token | null {
    return null;
  }

  public equals(obj: unknown): boolean {
    return isEqual(this, obj);
  }
}

export type TokenParser<T extends Token> = (buffer: CodePointBuffer, ...args: any[]) => T | null;
export type TokenConstructor<T extends Token> = new (...args: any[]) => T;

/**
 * A class that represents a boolean token.
 * @implements {Token}
 */
export class BooleanT extends Token {
  /**
   * Creates a new instance of `BooleanT`.
   * @param {string} value - The string representation of the boolean value.
   * @param {boolean} bool - The boolean value.
   */
  constructor(public value: string, public bool: boolean) {
    super();
  }

  public static parse(buffer: CodePointBuffer): BooleanT | null {
    buffer.savepoint();

    const s = buffer.readChunk();

    if (s.toLowerCase() === "true") return buffer.commitAndReturn(new BooleanT(s, true));
    else if (s.toLowerCase() === "false") return buffer.commitAndReturn(new BooleanT(s, false));
    else return buffer.rollbackAndReturn(null);
  }
}

// Skip CommandLine

export class DecimalPoint extends Token {
  public static parse(buffer: CodePointBuffer): DecimalPoint | null {
    return doParseCodepoint(buffer, ["."], DecimalPoint);
  }

  readonly value = ".";
}

export class Digit extends Token {
  constructor(public value: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): Digit | null {
    return doParseCodepoint(buffer, ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"], Digit);
  }
}

export class Digit1To9 extends Token {
  constructor(public value: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): Digit1To9 | null {
    return doParseCodepoint(buffer, ["1", "2", "3", "4", "5", "6", "7", "8", "9"], Digit1To9);
  }
}

export class DoubleQuoteString extends Token {
  constructor(public value: string, public content: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): DoubleQuoteString | null {
    return doParseQuoteString(buffer, '"', DoubleQuoteString);
  }
}

// Contains the values of the fractal that starts with character '.' and the sequence of digit followed
export class Frac extends Token {
  /**
   * Creates a new instance of `Frac`.
   * @param {string} value - The string representation of the fractional value.
   */
  constructor(public value: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): Frac | null {
    buffer.savepoint();

    let rtn = "";
    const d = DecimalPoint.parse(buffer);
    if (!d) {
      return buffer.rollbackAndReturn(null);
    }
    rtn += d.value;

    let hasDigit = false;
    let d0t9;
    while ((d0t9 = Digit.parse(buffer)) !== null) {
      rtn += d0t9.value;
      hasDigit = true;
    }

    if (!hasDigit) {
      return buffer.rollbackAndReturn(null);
    }

    return buffer.commitAndReturn(new Frac(rtn));
  }
}

export class Int extends Token {
  constructor(public value: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): Int | null {
    buffer.savepoint();

    const z = Zero.parse(buffer);
    if (z) {
      return buffer.commitAndReturn(new Int(z.value));
    } else {
      const p = PositiveInt.parse(buffer);

      if (!p) {
        return buffer.rollbackAndReturn(null);
      }

      return buffer.commitAndReturn(new Int(p.value));
    }
  }
}

export class Minus extends Token {
  public static parse(buffer: CodePointBuffer): Minus | null {
    return doParseCodepoint(buffer, ["-"], Minus);
  }

  readonly value = "-";
}

export class NegativeInt extends Token {
  constructor(public value: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): NegativeInt | null {
    buffer.savepoint();

    let rtn = "";
    const m = Minus.parse(buffer);
    if (!m) {
      return buffer.rollbackAndReturn(null);
    }
    rtn += m.value;

    const p = PositiveInt.parse(buffer);
    if (!p) {
      return buffer.rollbackAndReturn(null);
    }
    rtn += p.value;

    return buffer.commitAndReturn(new NegativeInt(rtn));
  }
}

export class NumberT extends Token {
  constructor(public value: string, public isPositive: boolean, public isDouble: boolean) {
    super();
  }

  public static parse(buffer: CodePointBuffer): NumberT | null {
    buffer.savepoint();

    let rtn = "";
    let isPositive: boolean, isDouble: boolean;

    const n = NegativeInt.parse(buffer);
    if (n) {
      rtn += n.value;
      isPositive = false;
    } else {
      const p = Int.parse(buffer);
      if (!p) {
        return buffer.rollbackAndReturn(null);
      }
      rtn += p.value;
      isPositive = true;
    }

    const f = Frac.parse(buffer);
    isDouble = f !== null;
    if (isDouble) {
      rtn += f!.value;
    }

    return buffer.commitAndReturn(new NumberT(rtn, isPositive, isDouble));
  }

  public toInt(): number {
    return parseInt(this.value);
  }

  public toDouble(): number {
    return parseFloat(this.value);
  }
}

// Skip OneLineInput

// Skip Parameter

export class PositiveInt extends Token {
  constructor(public value: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): PositiveInt | null {
    buffer.savepoint();

    let rtn = "";

    const d1t9 = Digit1To9.parse(buffer);
    if (!d1t9) {
      return buffer.rollbackAndReturn(null);
    }
    rtn += d1t9.value;

    let d0t9;
    while ((d0t9 = Digit.parse(buffer)) !== null) {
      rtn += d0t9.value;
    }

    return buffer.commitAndReturn(new PositiveInt(rtn));
  }
}

// Skip SafeString

export class SingleQuoteString extends Token {
  constructor(public value: string, public content: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): SingleQuoteString | null {
    return doParseQuoteString(buffer, "'", SingleQuoteString);
  }
}

export class StringT extends Token {
  constructor(public content: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): StringT | null {
    const c = buffer.peek();

    if (!c) {
      return null;
    } else if (c === '"') {
      const d = DoubleQuoteString.parse(buffer);
      if (!d) {
        return null;
      }
      return new StringT(d.content);
    } else if (c === "'") {
      const s = SingleQuoteString.parse(buffer);
      if (!s) {
        return null;
      }
      return new StringT(s.content);
    } else {
      const content = buffer.readChunk();
      return new StringT(content);
    }
  }
}

export class Zero extends Token {
  constructor(public value: string) {
    super();
  }

  public static parse(buffer: CodePointBuffer): Zero | null {
    return doParseCodepoint(buffer, ["0"], Zero);
  }
}

export class Parameter extends Token {
  private bool: BooleanT | null = null;
  private number: NumberT | null = null;
  private string: StringT | null = null;

  constructor(bool: BooleanT | null = null, number: NumberT | null = null, string: StringT | null = null) {
    super();
    this.bool = bool;
    this.number = number;
    this.string = string;
  }

  static parse(buffer: CodePointBuffer): Parameter | null {
    buffer.savepoint();

    let i = 0;
    while (!isDelimiter(buffer.peek(i))) i++;

    let firstStop = buffer.getIndex() + i;
    let bool = BooleanT.parse(buffer);
    if (bool != null) return buffer.commitAndReturn(new Parameter(bool));

    let number = NumberT.parse(buffer);
    if (number != null) {
      if (buffer.getIndex() == firstStop) return buffer.commitAndReturn(new Parameter(null, number));
      else {
        buffer.rollback();
        buffer.savepoint();
      }
    }

    let string = StringT.parse(buffer);
    if (string != null) return buffer.commitAndReturn(new Parameter(null, null, string));
    else return buffer.rollbackAndReturn(null);
  }

  getBoolean(): boolean {
    return this.bool!.bool;
  }

  getInt(): number {
    return this.number!.toInt();
  }

  getDouble(): number {
    return this.number!.toDouble();
  }

  getString(): string {
    return this.string!.content;
  }

  isBoolean(): boolean {
    return this.bool != null;
  }

  isNumber(): boolean {
    return this.number != null;
  }

  isDouble(): boolean {
    return this.number != null && this.number.isDouble;
  }

  isString(): boolean {
    return this.string != null;
  }

  toString(): string {
    if (this.isBoolean()) return this.bool!.bool + "";
    else if (this.isNumber()) return this.isDouble() ? this.number!.toDouble() + "" : this.number!.toInt() + "";
    else return this.getString();
  }

  equals(obj: unknown): boolean {
    if (obj === null) {
      return false;
    }

    if (!(obj instanceof Parameter)) {
      return false;
    }

    let other = obj as Parameter;

    return isEqual(other.string, this.string) && isEqual(other.number, this.number) && isEqual(other.bool, this.bool);
  }
}

/**
 * this record is a data class that contains the prefix and the parameters of a
 * command line
 */
export class CommandLine extends Token {
  /**
   * a static utility function that parses the CodePointBuffer, which is a wrapper
   * of a string, and returns a CommandLine object, which is a wrapper that
   * separated from the raw CodePointBuffer to two stuff, a command prefix and the
   * params right after the command prefix.
   *
   * via this process, it could assist the program to distinguish each commands
   * and tell the differences.
   *
   * @param buffer the CodePointBuffer to be parsed
   * @return the parsed CommandLine object, or null if the parse failed
   */

  constructor(public name: String, public params: Parameter[]) {
    super();
  }

  static parse(buffer: CodePointBuffer): CommandLine | null {
    buffer.readDelimiter();

    const cname: string = buffer.readChunk();
    if (cname.length === 0) {
      return null;
    }

    let params: Parameter[] = [];
    while (true) {
      buffer.readDelimiter();
      let p: Parameter | null = Parameter.parse(buffer);
      if (p === null) {
        break;
      }
      params.push(p);
    }
    if (buffer.hasNext()) {
      return null;
    } else {
      return new CommandLine(cname, params);
    }
  }
}
