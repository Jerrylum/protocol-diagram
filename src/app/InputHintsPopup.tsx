import { Box, Typography } from "@mui/material";
import { observer } from "mobx-react-lite";
import React from "react";
import { isFirefox } from "../core/Util";
import { Vector } from "../core/Vector";
import { BottomPanelController } from "./BottomPanel";

// Adopted from https://jsfiddle.net/dandv/aFPA7

// The properties that we copy into a mirrored div.
// Note that some browsers, such as Firefox,
// do not concatenate properties, i.e. padding-top, bottom etc. -> padding,
// so we have to do every single property specifically.
const MIRROR_PROPERTIES = [
  "boxSizing",
  "width", // on Chrome and IE, exclude the scrollbar, so the mirror div wraps exactly as the textarea does
  "height",
  "overflowX",
  "overflowY", // copy the scrollbar for IE

  "borderTopWidth",
  "borderRightWidth",
  "borderBottomWidth",
  "borderLeftWidth",

  "paddingTop",
  "paddingRight",
  "paddingBottom",
  "paddingLeft",

  // https://developer.mozilla.org/en-US/docs/Web/CSS/font
  "fontStyle",
  "fontVariant",
  "fontWeight",
  "fontStretch",
  "fontSize",
  "lineHeight",
  "fontFamily",

  "textAlign",
  "textTransform",
  "textIndent",
  "textDecoration", // might not make a difference, but better be safe

  "letterSpacing",
  "wordSpacing"
] as const;

export function getMirrorDiv(reference: HTMLInputElement | HTMLTextAreaElement): HTMLDivElement {
  const ownerDocument = reference.ownerDocument;

  const mirror = ownerDocument.createElement("div") as HTMLDivElement;
  ownerDocument.body.appendChild(mirror);

  const mirrorStyle = mirror.style;
  const referenceStyle = getComputedStyle(reference);

  // default textarea styles
  mirrorStyle.whiteSpace = "pre-wrap";
  if (reference.nodeName !== "INPUT") mirrorStyle.wordWrap = "break-word"; // only for textarea-s

  // position off-screen
  mirrorStyle.position = "absolute"; // required to return coordinates properly
  mirrorStyle.visibility = "hidden"; // not 'display: none' because we want rendering

  // transfer the element's properties to the div
  MIRROR_PROPERTIES.forEach(prop => ((mirrorStyle as any)[prop] = (referenceStyle as any)[prop]));

  if (isFirefox()) {
    mirrorStyle.width = parseInt(referenceStyle.width) - 2 + "px"; // Firefox adds 2 pixels to the padding - https://bugzilla.mozilla.org/show_bug.cgi?id=753662
    // Firefox lies about the overflow property for textareas: https://bugzilla.mozilla.org/show_bug.cgi?id=984275
    if (reference.scrollHeight > parseInt(referenceStyle.height)) mirrorStyle.overflowY = "scroll";
  } else {
    mirrorStyle.overflow = "hidden"; // for Chrome to not render a scrollbar; IE keeps overflowY = 'scroll'
  }

  return mirror as HTMLDivElement;
}

export function getCaretCoordinates(reference: HTMLInputElement | HTMLTextAreaElement, position: number) {
  const mirrorDiv = getMirrorDiv(reference);

  const computed = getComputedStyle(reference);

  mirrorDiv.textContent = reference.value.substring(0, position);
  // the second special handling for input type="text" vs textarea: spaces need to be replaced with non-breaking spaces - http://stackoverflow.com/a/13402035/1269037
  if (reference.nodeName === "INPUT") mirrorDiv.textContent = mirrorDiv.textContent.replace(/\s/g, "\u00a0");

  var span = document.createElement("span");
  // Wrapping must be replicated *exactly*, including when a long word gets
  // onto the next line, with whitespace at the end of the line before (#7).
  // The  *only* reliable way to do that is to copy the *entire* rest of the
  // textarea's content into the <span> created at the caret position.
  // for inputs, just '.' would be enough, but why bother?
  span.textContent = reference.value.substring(position) || "."; // || because a completely empty faux span doesn't render at all
  span.style.backgroundColor = "lightgrey";
  mirrorDiv.appendChild(span);

  const coordinates = new Vector(
    span.offsetLeft + parseInt(computed["borderLeftWidth"]),
    span.offsetTop + parseInt(computed["borderTopWidth"])
  );

  mirrorDiv.remove();

  return coordinates;
}

export type AcceptedValueProps = {
  controller: BottomPanelController;
  value: string;
  current: string;
  selected: boolean;
};

export const AcceptedValue = observer((props: AcceptedValueProps) => {
  const [isHover, setIsHover] = React.useState(false);

  let bgColor: string = "transparent";
  if (props.selected && isHover) bgColor = "rgb(25 118 210 / 70%)";
  if (props.selected && !isHover) bgColor = "rgb(25 118 210 / 60%)";
  if (!props.selected && isHover) bgColor = "rgb(235 235 235 / 100%)";

  const head = props.value.substring(0, props.current.length);
  const tail = props.value.substring(props.current.length);

  return (
    <Box
      onMouseEnter={() => setIsHover(true)}
      onMouseLeave={() => setIsHover(false)}
      onClick={() => {
        props.controller.insertAutoCompletionValue(props.value);
        props.controller.inputElement?.focus();
      }}
      sx={{
        padding: "0 4px",
        backgroundColor: bgColor,
        cursor: "pointer"
      }}>
      {head !== "" && (
        <Typography
          variant="body2"
          sx={{
            fontFamily: "Ubuntu Mono",
            color: "text.primary",
            minHeight: "20px",
            lineHeight: "20px",
            display: "inline-block",
            fontWeight: 700
          }}>
          {head}
        </Typography>
      )}
      {tail !== "" && (
        <Typography
          variant="body2"
          sx={{
            fontFamily: "Ubuntu Mono",
            color: "text.primary",
            minHeight: "20px",
            lineHeight: "20px",
            display: "inline-block"
          }}>
          {tail}
        </Typography>
      )}
    </Box>
  );
});

export const InputHintsPopup = observer((props: { controller: BottomPanelController }) => {
  const controller = props.controller;

  const mapping = controller.mapping;
  if (mapping === null) return null;

  const spec = mapping.spec;
  if (spec === null) return null;

  const currentParamValue = mapping.param?.value.value ?? "";
  const autoCompletionValues = controller.autoCompletionValues;

  const caretOffset = controller.inputElement ? getCaretCoordinates(controller.inputElement, mapping.startIndex).x : 0;

  return (
    <Box
      onMouseDown={() => (controller.isFocusedPopup = true)}
      onMouseUp={() => (controller.isFocusedPopup = false)}
      sx={{
        position: "absolute",
        left: caretOffset + "px",
        bottom: "calc(100% - 8px)",
        minWidth: "100px",
        backgroundColor: "rgb(250, 250, 250)",
        userSelect: "none"
      }}>
      {autoCompletionValues.length > 0 ? (
        <Box
          sx={{
            bottom: "calc(100% - 8px)",
            // maxHeight: "120px",
            // overflowY: "auto",
            minWidth: "100px",
            border: "1px solid rgba(0, 0, 0, 0.23)",
            borderBottom: "none"
          }}>
          {autoCompletionValues.map((val, i) => (
            <AcceptedValue
              key={i}
              controller={controller}
              value={val}
              current={currentParamValue}
              selected={controller.selected === val}
            />
          ))}
        </Box>
      ) : null}

      <Typography
        variant="body2"
        sx={{
          fontFamily: "Ubuntu Mono",
          color: "text.primary",
          minHeight: "20px",
          lineHeight: "20px",
          padding: "0 4px",
          border: "1px solid rgba(0, 0, 0, 0.23)"
        }}>
        {spec.description}
      </Typography>
    </Box>
  );
});

