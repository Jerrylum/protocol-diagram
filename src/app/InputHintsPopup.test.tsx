import { InputHintsPopup, getMirrorDiv, getCaretCoordinates } from "./InputHintsPopup";
import { fireEvent, render } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { BottomPanelController } from "./BottomPanel";
import { cpb } from "../token/Tokens.test";
import { BooleanT, CommandParameter } from "../token/Tokens";
import {
  ClearCommand,
  ConfigCommand,
  buildInputSpecByCommands,
  buildInputSpecByUsages,
  mapCommandParameterWithInputSpec
} from "../command/Commands";

test("getMirrorDiv", () => {
  Object.defineProperty(HTMLTextAreaElement.prototype, "scrollHeight", {
    configurable: true,
    get: function () {
      return this._scrollHeight || 0;
    },
    set(val) {
      this._scrollHeight = val;
    }
  });

  const htmlInputArea = document.createElement("input");
  htmlInputArea.value = "test";
  const mirrorDiv = getMirrorDiv(htmlInputArea);
  expect(mirrorDiv).toMatchSnapshot();

  (window as any)["mozInnerScreenX"] = "mocktest";
  const components = (
    <div id="root-container">
      <textarea id={"test"} style={{ height: 1, width: 10, overflowY: "scroll" }} />
    </div>
  );
  const result = render(components);
  const textField = result.container.querySelector("#test") as HTMLTextAreaElement;
  (textField as any).scrollHeight = 100;

  result.rerender(components);
  const mirrorDiv2 = getMirrorDiv(textField);
  expect(mirrorDiv2).toMatchSnapshot();
});

test("Render InputHintsPopup", () => {
  const btmpController = new BottomPanelController();
  const components = (
    <div id="root-container">
      <InputHintsPopup controller={btmpController} />
    </div>
  );
  const result = render(components);

  const inspec = buildInputSpecByUsages([
    {
      name: "testbool",
      paramType: BooleanT,
      description: "test description"
    },
    {
      name: "testbool2",
      paramType: BooleanT,
      description: "test description"
    }
  ]);
  const commandParameters = [CommandParameter.parse(cpb("true"))!];
  const map = mapCommandParameterWithInputSpec(commandParameters, null);
  btmpController.mapping = map[0];
  result.rerender(components);

  const map2 = mapCommandParameterWithInputSpec(commandParameters, inspec!);
  btmpController.mapping = map2[0];
  result.rerender(components);

  const map3 = mapCommandParameterWithInputSpec(commandParameters, inspec!);
  btmpController.mapping = map3[1];
  btmpController.inputElement = document.createElement("input");
  result.rerender(components);

  const commandParameters2 = [CommandParameter.parse(cpb("c"))!];

  const map4 = mapCommandParameterWithInputSpec(
    commandParameters2,
    buildInputSpecByCommands([new ClearCommand(), new ConfigCommand()])
  );
  btmpController.mapping = map4[0];
  result.rerender(components);
  expect(result.container).toMatchSnapshot();

  const inputHintsPopupBox = result.container.querySelector("#root-container > div");
  act(() => {
    fireEvent.mouseDown(inputHintsPopupBox!);
  });
  result.rerender(components);
  expect(btmpController.isFocusedPopup).toBe(true);
  act(() => {
    fireEvent.mouseUp(inputHintsPopupBox!);
  });
  result.rerender(components);
  expect(btmpController.isFocusedPopup).toBe(false);

  const clearCommandBox = result.container.querySelectorAll("#root-container > div > div > div")[0];
  const configCommandBox = result.container.querySelectorAll("#root-container > div > div > div")[1];
  act(() => {
    fireEvent.mouseEnter(clearCommandBox);
  });
  result.rerender(components);
  act(() => {
    fireEvent.mouseLeave(clearCommandBox);
  });
  result.rerender(components);
  act(() => {
    fireEvent.mouseEnter(configCommandBox);
  });
  result.rerender(components);
  act(() => {
    fireEvent.mouseLeave(configCommandBox);
  });
  result.rerender(components);
  act(() => {
    fireEvent.click(clearCommandBox);
  });
  result.rerender(components);
  expect(btmpController.inputElement.value).toBe("clear ");
});
