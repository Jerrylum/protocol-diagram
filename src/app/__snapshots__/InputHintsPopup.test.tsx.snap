// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render InputHintsPopup 1`] = `
<div>
  <div
    id="root-container"
  >
    <div
      class="MuiBox-root css-1sgq4p7"
    >
      <div
        class="MuiBox-root css-a35xdc"
      >
        <div
          class="MuiBox-root css-13w71a5"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-oldu9d-MuiTypography-root"
          >
            c
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-1kvvwtc-MuiTypography-root"
          >
            lear
          </p>
        </div>
        <div
          class="MuiBox-root css-m7mt60"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-oldu9d-MuiTypography-root"
          >
            c
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-1kvvwtc-MuiTypography-root"
          >
            onfig
          </p>
        </div>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 css-1gkecma-MuiTypography-root"
      >
        the command to execute
      </p>
    </div>
  </div>
</div>
`;

exports[`getMirrorDiv 1`] = `
<div
  style="white-space: pre-wrap; position: absolute; visibility: hidden; border-top-width: 2px; border-right-width: 2px; border-bottom-width: 2px; border-left-width: 2px; padding: 1px 1px 1px 1px; line-height: normal; font-family: -webkit-small-control; text-align: start; text-transform: none; text-indent: 0; letter-spacing: normal; word-spacing: normal; overflow: hidden;"
/>
`;

exports[`getMirrorDiv 2`] = `
<div
  style="white-space: pre-wrap; word-wrap: break-word; position: absolute; visibility: hidden; width: 8px; height: 1px; overflow-y: scroll; border-top-width: 1px; border-right-width: 1px; border-bottom-width: 1px; border-left-width: 1px; padding: 2px 2px 2px 2px; line-height: normal; font-family: -webkit-small-control; text-align: start; text-transform: none; text-indent: 0; letter-spacing: normal; word-spacing: normal;"
/>
`;
